File: singly-linked-lists/singly-linked-list.go
// main package implements Singly Linked List using Golang.
package main

import "fmt"

// Node struct creates a new node for SLL.
type Node struct {
	value int
	next  *Node
}

// Next method points to the next node linked to a given node.
func (n *Node) Next() *Node {
	return n.next
}

// List struct creates a type for SLL.
type List struct {
	head   *Node
	tail   *Node
	length int
}

// First method returns the head or the first element of the SLL.
func (l *List) First() *Node {
	return l.head
}

// Last methos returns the tail or the last element of the SLL.
func (l *List) Last() *Node {
	return l.tail
}

// Traverse method prints the values of each node of the SLL.
func (l *List) Traverse() {
	n := l.First()

	for {
		fmt.Println(n.value)
		n = n.Next()
		if n == nil {
			break
		}
	}
}

// Push method takes a value as argument and appends a node at the end of the list with the given value.
func (l *List) Push(val int) *List {
	newNode := &Node{value: val}
	if l.head == nil {
		l.head = newNode
	} else {
		l.tail.next = newNode
	}
	l.tail = newNode
	l.length++
	return l
}

// Pop method returns the tail or the last node of the list.
func (l *List) Pop() *Node {
	if l.length == 0 {
		return nil
	}
	current := l.First()
	newTail := current
	for current.next != nil {
		newTail = current
		current = current.next
	}
	l.tail = newTail
	l.tail.next = nil
	if l.length == 0 {
		l.head = nil
		l.tail = nil
	}
	l.length--
	return current
}

// Shift method removes and returns the first node of the SLL.
func (l *List) Shift() *Node {
	if l.head == nil {
		return nil
	}
	currentHead := l.First()
	l.head = currentHead.next
	l.length--
	if l.length == 0 {
		l.head = nil
		l.tail = nil
	}
	return currentHead
}

// Unshift takes a value as parameter and adds a node with the given value at the beginning of the SLL.
func (l *List) Unshift(value int) *List {
	newNode := &Node{value: value}
	if l.First() == nil {
		l.head = newNode
		l.tail = l.head
	} else {
		newNode.next = l.head
		l.head = newNode
	}
	l.length++
	return l
}

// Get method takes an index and return the node at the index or nil.
func (l *List) Get(idx int) *Node {
	if idx < 0 || idx > l.length {
		return nil
	}
	currentNode := l.head
	currentIdx := 0
	for idx != currentIdx {
		currentNode = currentNode.next
		currentIdx++
	}
	return currentNode
}

// Set method, takes an index and a value as arguments and replace the value of the node at the given index.
func (l *List) Set(idx, val int) bool {
	foundNode := l.Get(idx)
	if foundNode != nil {
		foundNode.value = val
		return true
	}
	return false
}

// Insert method, takes an index and value as argumetns and inserts a new node at the give index.
func (l *List) Insert(idx, val int) bool {
	if idx < 0 || idx > l.length {
		return false
	}
	if idx == l.length {
		l.Push(val)
		return true
	}
	if idx == 0 {
		l.Unshift(val)
		return true
	}
	newNode := &Node{value: val}
	prevNode := l.Get(val)
	nextNode := prevNode.next
	prevNode.next = newNode
	newNode.next = nextNode
	l.length++
	return true
}

// Remove method, takes an index as argument, removes the node at the index and returns the node.
func (l *List) Remove(idx int) *Node {
	if idx < 0 || idx >= l.length {
		return nil
	}
	if idx == 0 {
		return l.Shift()
	}
	if idx == l.length-1 {
		return l.Pop()
	}
	beforeNode := l.Get(idx - 1)
	removedNode := beforeNode.next
	beforeNode.next = removedNode.next
	l.length--
	return removedNode
}

func main() {
	sll := &List{}
	sll.Push(0)
	sll.Push(1)
	sll.Push(2)
	sll.Push(3)
	sll.Push(4)
	fmt.Println("Initial list")
	sll.Traverse()
	fmt.Println("List length", sll.length)
	fmt.Println("----------------------------")
	fmt.Println("Removed node: ", sll.Remove(3))
	fmt.Println("----------------------------")
	fmt.Println("Updated list")
	sll.Traverse()
	fmt.Println("List length", sll.length)
}

File: singly-linked-lists/singly-linked-list.js
// Singly Likned List with Java Script implementation.

// a class that represents a node of the Singly Linked List.
class Node {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

// SLL class.
class SinglyLinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    // Traverse method, prints all the nodes in the SLL.
    Traverse() {
        var currentNode = this.head;
        if (this.length === 0) return null;
        while (currentNode != null) {
            console.log(currentNode.val)
            currentNode = currentNode.next;
        }
    }
    // Push method takes a value as argumant and appends it at the end of the list.
    Push(val) {
        var newNode = new Node(val);
        if (!this.head) {
            this.head = newNode;
            this.tail = this.head;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
        return this;
    }

    // Pop method removes the last node of the list and returns it.
    Pop() {
        if (this.length === 0) {
            return undefined
        }
        var current = this.head;
        var newTail = current
        while (current.next) {
            newTail = current;
            current = current.next;
        }
        this.tail = newTail;
        this.tail.next = null;
        this.length--;
        if (this.length === 0) {
            this.head = null;
            this.tail = null;
        }
        return current;
    }

    // Shift method removes the first node of the list and returns it.
    Shift() {
        if (this.length === 0) {
            return undefined;
        }

        var currentHead = this.head;
        this.head = currentHead.next;
        this.length--;
        if (this.length === 0) {
            this.head = null;
            this.tail = null;
        }
        return currentHead;
    }

    // Unshift method takes a value as argument, and inserts a new node at the beginning of the list with the input value.
    Unshift(val) {
        var newNode = new Node(val);
        if (!this.head) {
            this.head = newNode;
            this.tail = this.head;
        } else {
            newNode.next = this.head;
            this.head = newNode;
        }
        this.length++;
        return this;
    }

    // Get method accepts an index and return the node at that index or null if there is no node at that index.
    Get(idx) {
        if (idx < 0 || idx > this.length) {
            return null
        }
        var currentNode = this.head;
        var currentIdx = 0;

        while (idx != currentIdx) {

            currentNode = currentNode.next;
            currentIdx++;
        }
        return currentNode
    }

    // Set method, takes an index and a value as arguments and replace the value of the node at the given index.
    Set(idx, val) {
        var foundNode = this.Get(idx);
        if (foundNode) {
            foundNode.val = val;
            return true;
        }
        return false;
    }

    // Insert method, takes an index and value as arguments and inserts a new node at the give index.
    Insert(idx, val) {
        if (idx < 0 || idx > this.length) return false;
        if (idx === this.length) return !!this.Push(val);
        if (idx === 0) return !!this.Unshift(val);
        var newNode = new Node(val);
        var prevNode = this.Get(idx - 1);
        var nextNode = prevNode.next;
        prevNode.next = newNode;
        newNode.next = nextNode;
        this.length++;
        return true;
    }
    // Remove method, takes an index as argument, removes the node at the index and returns the node.
    Remove(idx) {
        if (idx < 0 || idx >= this.length) return undefined;
        if (idx === this.length - 1) return this.Pop();
        if (idx === 0) return this.Shift();
        var beforeNode = this.Get(idx - 1);
        var removedNode = beforeNode.next;
        beforeNode.Next = removedNode.next;
        this.lendth--;
        return removedNode;
    }

    // Reverse methos, reverses the SLL.
    Reverse() {
        var node = this.head;
        this.head = this.tail;
        this.tail = this.head;
        var nextNode;
        var prevNode = null;
        for (let i = 0; i < this.length; i++) {
            nextNode = node.next;
            node.next = prevNode;
            prevNode = node;
            node = nextNode;
        }
        return this;
    }
}

let list = new SinglyLinkedList();
list.Push(0);
list.Push(1);
list.Push(2);
list.Push(3);
list.Push(4);
console.log("Initial list", list.Traverse());
console.log("*********************************************")
console.log("Removed Node", list.Reverse())
console.log("*********************************************")
console.log("Updated list", list.Traverse());